---
title: "Meta Populations"
author: "Ted Monyak"
date: "2025-01-16"
output: html_document
description: This notebook contains scripts for developing breeding metapopulations,
creating mapping populations, and running linkage mapping and GWAS
---

```{r setup, include=FALSE, echo=FALSE}
require("knitr")

rm(list = ls())

# Set to false if running on Alpine
runLocal = TRUE

if (runLocal) {
  opts_knit$set(root.dir = "~/Documents/CSU/R/BreedingSims")
} else {
  opts_knit$set(root.dir = "/pl/active/Morris_CSU/Ted_Monyak/BreedingSims")
}
```

Imports
```{r}
library(AlphaSimR)
library(ade4)
library(devtools)
library(dplyr)
library(ggpubr)
library(ggplot2)
library(factoextra)
library(patchwork)
library(plotly)
library(purrr)
library(qtl)
library(qtlTools)
library(reshape2)
library(rrBLUP)
library(snow)
library(tibble)
library(tidyr)
library(viridis)

set.seed(123)

if (runLocal) {
  setwd("~/Documents/CSU/R/BreedingSims")
  output_dir <- file.path(getwd(), "Output")
} else {
  setwd("/pl/active/Morris_CSU/Ted_Monyak/BreedingSims")
  output_dir <- ("/scratch/alpine/c837220672@colostate.edu/Output")
}

if (!dir.exists(output_dir)) dir.create(output_dir)

source("Functions/Fitness.R")
source("Functions/GenoConversions.R")
source("Functions/MappingPopulations.R")
source("Functions/PopGen.R")
source("Functions/QtlMapping.R")
source("Functions/TraitArchitecture.R")
source("Scripts/GlobalParameters.R")
```

Show that RILS generated from trans-elite crosses of independently adapting populations have a wider distribution of trait values than the original elite populations.
```{r}
# Tidy dataframe to store the results
result_df <- data.frame(sim=rep(1:n.sims, each=3),
                        traitVal1=numeric(n.sims*3),
                        traitVal2=numeric(n.sims*3),
                        pop=character(n.sims*3))

# Run n.sims simulations
s <-1
repeat {
  print(paste0("Sim: ", (s+2)/3))
  # Each value of s is used for 3 rows in the dataframe. Terminate once s*3 is reached
  if (s > n.sims*3) {
    break
  }
  pop <- founderPop

  # Burn-in
  for (gen in 1:n.burnInGens) {
    pop <- selectCross(pop, trait=twoTraitFitFunc, nInd=nInd(pop)*n.burnInSelProp, nCrosses=nInd(pop))
  }

  # Randomly select two subpopulations
  popA <- selectInd(pop, use="rand", nInd=n.subPopSize)
  popB <- selectInd(pop, use="rand", nInd=n.subPopSize)

  # Move the two subpopulations towards a fitness optimum
  for (gen in 1:n.gens) {
    if (mean(twoTraitFitFunc(pheno(popA))) < n.margin) {
      # Calculate selection ratio as a function of current fitness
      selRat <- selectionRatio(abs(mean(twoTraitFitFunc(pheno(popA)))))
      popA <- selectCross(popA, trait=twoTraitFitFunc, nInd=nInd(popA)*selRat, nCrosses=nInd(popA))
    }
    if (mean(twoTraitFitFunc(pheno(popB))) < n.margin) {
      selRat <- selectionRatio(abs(mean(twoTraitFitFunc(pheno(popB)))))
      popB <- selectCross(popB, trait=twoTraitFitFunc, nInd=nInd(popB)*selRat, nCrosses=nInd(popB))
    }
  }

  # Randomly select a parent from each subpopulation
  parentA <- popA[runif(1,1,nInd(popA))]
  parentB <- popB[runif(1,1,nInd(popB))]
  
  # Cross the parents to create a biparental RIL population with n.RILFams families
  F1 <- randCross2(parentA,
                   parentB,
                   nCrosses=1,
                   nProgeny=n.RILFams)

  F2 <- self(F1)
  F3 <- self(F2)
  F4 <- self(F3)
  F5 <- self(F4)
  F6 <- self(F5)
  F7 <- self(F6)
  F8 <- self(F7)
  F9 <- self(F8)

  # Create F10s with n.indPerRILFam replicates per family
  RIL <- self(F9, nProgeny=n.indPerRILFam)

  # Update result dataframes with mean phenotype values from each population
  result_df$traitVal1[s] <- meanP(popA)[1]
  result_df$traitVal2[s] <- meanP(popA)[2]
  result_df$pop[s] <- "Elite Pop 1"
  # Increment index
  s <- s+1
  result_df$traitVal1[s] <- meanP(popB)[1]
  result_df$traitVal2[s] <- meanP(popB)[2]
  result_df$pop[s] <- "Elite Pop 2"
  # Increment index
  s <- s+1
  result_df$traitVal1[s] <- meanP(RIL)[1]
  result_df$traitVal2[s] <- meanP(RIL)[2]
  result_df$pop[s] <- "RIL"
  s <- s+1
}

# Create plot for first trait value
t1 <- ggplot(data=result_df, aes(x=traitVal1, fill=pop)) +
  geom_histogram(position='identity', binwidth = 0.05, alpha=0.6) +
  scale_fill_manual(name = "Population",
                      values = c("RIL" = "darkblue",
                                "Elite Pop 1" = "lightblue",
                                "Elite Pop 2" = "lightpink")) +
  labs(title="Trait 1")

# Create plot for second trait value
t2 <- ggplot(data=result_df, aes(x=traitVal2, fill=pop)) +
  geom_histogram(position='identity', binwidth = 0.05, alpha=0.6) +
  scale_fill_manual(name = "Population",
                    values = c("RIL" = "darkblue",
                               "Elite Pop 1" = "lightblue",
                               "Elite Pop 2" = "lightpink")) +
  labs(title="Trait 2")

g <- (t1|t2)

save_dir <- file.path(output_dir, "EliteVSRIL")
if (!dir.exists(save_dir)) dir.create(save_dir)
save_dir <- file.path(save_dir, format(Sys.time(), "%F_%H_%M_%S"))
if (!dir.exists(save_dir)) dir.create(save_dir)
write.table(getParams(), file.path(save_dir, "params.txt"), col.names=FALSE, quote=FALSE, sep=":\t")

fname <- file.path(save_dir, "EliteVSRil.pdf")
ggplot2::ggsave(filename = fname,
                device = "pdf")
```

1) Pull out individuals from each population and cross them
2) Create RILs
3) Do QTL Mapping
```{r}
save_dir <- file.path(output_dir, "QtlMapping")
if (!dir.exists(save_dir)) dir.create(save_dir)
save_dir <- file.path(save_dir, format(Sys.time(), "%F_%H_%M_%S"))
dir.create(save_dir)

source("Scripts/GlobalParameters.R")
source("Scripts/CreateFounderPop.R")
eff_size.df <- data.frame(orderFixed=c(),
                          effectSize=c())
saveAllelePlots <- TRUE
saveQtlPlots <- TRUE
source("Scripts/CreateIndependentPops.R")

# Increase heritability for the breeding portion of the simuation
SP$setVarE(h2=c(n.h2Breeding, n.h2Breeding))
# Create a biparental RIL family
# pops is created in CreateIndependentPops
res <- createRIL(popA=pops[[1]], popB=pops[[2]], save_dir=save_dir, inter=TRUE)
# Parents A and B are the first two individuals in the result population
RIL <- res[-(1:2)]
parentA <- res[1]
parentB <- res[2]

print(getSigQtl(RIL, parentA, parentB, save_dir))
```

```{r}
save_dir <- file.path(output_dir, "Yield")
if (!dir.exists(save_dir)) dir.create(save_dir)
save_dir <- file.path(save_dir, format(Sys.time(), "%F_%H_%M_%S"))
dir.create(save_dir)

source("Scripts/GlobalParameters.R")
source("Scripts/CreateFounderPop.R")
saveFitnessPlots <- TRUE
n.yieldCor <- -1
n.yieldProp <- 0.1

w <- twoTraitFitFunc(pheno(founderPop[1]))
pheno(founderPop[1])[3]

w <- -0.08
y <- -0.02

w*y * n.yieldCor * -1

source("Scripts/CreateIndependentPops.R")
cor(pheno(pop), pheno(pop))

twoTraitFitFunc(pheno(pop[1]))
threeTraitFitFunc(pheno(pop[1]))
pheno(pop[1])
meanP(pop)
```

Add effect sizes to LOD plots (work-in-progress)
```{r}
cross <- getCross(RIL, parentA, parentB, "riself")
cross <- drop.nullmarkers(cross)
plot(cross)

plot.map(pull.map(cross),est.map(cross))
phes <- phenames(cross)[1:2]
cross <- calc.genoprob(cross, step=n.step, error.prob=n.errorProb)
out.hk <- scanone(cross, pheno.col=phes, method=n.mappingMethod,
                  n.cluster=n.cores)
operm.hk <- scanone(cross, pheno.col=phes, method=n.mappingMethod,
                    n.perm=1000,n.cluster=n.cores)

```

Figure out difference (inter-intra) in mean qtl per population
```{r}
res.df %>%
  group_by(pop, type) %>%
  summarize(meanQtlSize=mean(nSigQtl), .groups='drop') %>%
  group_by(pop) %>%
  summarize(qtldiff=-diff(meanQtlSize)) %>%
  ggplot(aes(x=qtldiff)) + geom_bar(stat="bin", binwidth=0.5)
```

Create NAM and DP Populations  for GWAS
```{r}
n.nPops <- 11
n.subPopSize <- 80
source("Scripts/CreateIndependentPops.R")

FST(pops)
NAM <- createNAM()
DP <- createDP()

plotTraitArchitecture(NAM)
```


Run GWAS for 2 traits
```{r}
pop <- DP
 
# Get genotype in rrBLUP encoded format
geno <- getGwasGeno(pop)

# Get phenotype data
# dataframe has 3 columns: id, pheno.Trait1, pheno.Trait2
pheno = data.frame(id = 1:pop@nInd,
                  pheno  = pop@pheno)

# Run GWAS for 2 traits
save_dir <- file.path(output_dir, "GWAS")
if (!dir.exists(save_dir)) dir.create(save_dir)
save_dir <- file.path(save_dir, format(Sys.time(), "%F_%H_%M_%S"))
if (!dir.exists(save_dir)) dir.create(save_dir)
fname <- file.path(save_dir, "gwas_trait1_NAM.pdf")
pdf(fname)

# select the 2nd column for trait 1
model.1 <- GWAS(pheno[,c(1,2)], geno, plot = T, n.core=n.cores)
dev.off()

fname <- file.path(save_dir, "gwas_trait2_NAM.pdf")
pdf(fname)
# select the 3rd column for trait 2
model.2 <- GWAS(pheno[,c(1,3)], geno, plot = T, n.core=n.cores)
dev.off()


model.pc = GWAS(pheno[,c(1,trait+1)], geno, n.PC = 3, plot = T)

model1$Trait1 = 10 ^ (-model1$pheno.Trait1)
model2$Trait1 = 10 ^ (-model2$pheno.Trait1)

model1$Trait2 = 10 ^ (-model1$pheno.Trait2)
model2$Trait2 = 10 ^ (-model2$pheno.Trait2)

qtl1 = as.vector(getQtlMap(trait = 1)$id)
qtl2 = as.vector(getQtlMap(trait = 2)$id)
par(mfrow = c(2, 1))
manhattan(model1, chr = "chr", bp = "pos", p = "pheno.Trait2", snp = "snp", highlight = qtl2,
  main = "Marker", ylim = c(0,10))
manhattan(model2, chr = "chr", bp = "pos", p = "pheno.Trait2", snp = "snp", highlight = qtl2,
  main = "Marker + principal components", ylim = c(0,10))
# Check QQ-plot
par(mfrow = c(2, 1))
qq(model1$Trait1, main = "Marker")
qq(model2$Trait1, main = "Marker + principal components")
```

QTL stuff not being used
```{r}
plot(cross)
nmar(cross)
plot.map(pull.map(cross),est.map(cross))
#summary(operm.hk, alpha=0.05, pvalues=TRUE)


#mar <- find.marker(cross, chr=10, pos=63.3)
#plotPXG(cross, marker=mar)
#effectplot(cross, mname1=mar)
#out.cim <- cim(cross, n.marcovar=3)
#plot(out.cim)

data("multitrait")
data(multitrait)
data(map10)
plot(map10)
simcross <- sim.cross(map10, type="f2", n.ind=100, missing.prob=0.02)
plot.map(pull.map(simcross),est.map(simcross))
geno.table(simcross, chr=1)

cross <- est.rf(cross)
plotRF(cross)
plotMissing(cross)
geno.table(cross, chr=1)
cg <- comparegeno(cross)
hist(cg[lower.tri(cg)], breaks=seq(0, 1, len=101), xlab="No. matching genotypes")
print(dup <- findDupMarkers(cross, exact.only=FALSE))
rug(cg[lower.tri(cg)])
gt <- geno.table(cross)
gt[gt$P.value < 0.05/totmar(cross),]

g <- pull.geno(cross)
gfreq <- apply(g, 1, function(a) table(factor(a, levels=1:3)))
gfreq <- t(t(gfreq) / colSums(gfreq))
par(mfrow=c(1,3), las=1)
for(i in 1:3) {
  plot(gfreq[i,], ylab="Genotype frequency", main=c("AA", "AB", "BB")[i], ylim=c(0,1))
}

newMap <- est.map(cross, error.prob=0.01)
plotMap(cross, newMap)
```